homeassistant/components/tts/__init__.py

```diff
@@ -0,0 +1,370 @@
+"""
+Provide functionality to TTS.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/tts/
+"""
+import asyncio
+import logging
+import hashlib
+import mimetypes
+import os
+import re
+
+from aiohttp import web
+import voluptuous as vol
+
+from homeassistant.const import ATTR_ENTITY_ID
+from homeassistant.bootstrap import async_prepare_setup_platform
+from homeassistant.core import callback
+from homeassistant.config import load_yaml_config_file
+from homeassistant.components.http import HomeAssistantView
+from homeassistant.components.media_player import (
+    SERVICE_PLAY_MEDIA, MEDIA_TYPE_MUSIC, ATTR_MEDIA_CONTENT_ID,
+    ATTR_MEDIA_CONTENT_TYPE, DOMAIN as DOMAIN_MP)
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import config_per_platform
+import homeassistant.helpers.config_validation as cv
+
+DOMAIN = 'tts'
+DEPENDENCIES = ['http']
+
+_LOGGER = logging.getLogger(__name__)
+
+MEM_CACHE_TIME = 300
+MEM_CACHE_FILENAME = 'filename'
+MEM_CACHE_VOICE = 'voice'
+
+CONF_LANG = 'language'
+CONF_CACHE = 'cache'
+CONF_CACHE_DIR = 'cache_dir'
+
+DEFAULT_CACHE = True
+DEFAULT_CACHE_DIR = "tts"
+DEFAULT_LANG = 'en'
+
+SERVICE_SAY = 'say'
+
+ATTR_MESSAGE = 'message'
+ATTR_CACHE = 'cache'
+
+_RE_VOICE_FILE = re.compile(r"([a-f0-9]{40})_([a-z]+)\.[a-z0-9]{3,4}")
+
+PLATFORM_SCHEMA = cv.PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_LANG, default=DEFAULT_LANG): cv.string,
+    vol.Optional(CONF_CACHE, default=DEFAULT_CACHE): cv.boolean,
+    vol.Optional(CONF_CACHE_DIR, default=DEFAULT_CACHE_DIR): cv.string,
```

Comment:
I think that we should fully own the caching strategy instead of having users configure it.

 - Always cache
 - Cache all providers to the same dir. Have filenames contain provider.
 - Never store cache in memory
 - When loading cache from disk, also touch it so last modified time gets updated
 - Auto clean up the cache dir once a day, using file last modified time to remove anything that hasn't been used in 48h